### ISSUE TITLE
targets = batch[-1]

### ISSUE THREAD

**Issue by HBU-Lin-Li** [vturrisi/solo-learn#57]
Hi, thanks for your sharing.
What is the meaning of the batch？

Thanks

**Comment by HBU-Lin-Li** [comment]
I want to achieve 'NNCLR'，but i'm having some problems.

**Comment by DonkeyShot21** [comment]
Hi! Thanks for your interest in our repo!

The batch is a tuple (index, views, targets) so by doing batch[-1] you are selecting the targets. Note that the targets in nnclr are only used to compute the metrics, not for training.

What problems are you having exactly? Did you install the requirements properly as described in the readme? Have you tried using the commands in `/bash_files/pretrain`?

**Comment by HBU-Lin-Li** [comment]
Thank for your reply.
I want to  incorporate NN from NNCLR into something else.When i use the function of 'dequeue_and_enqueue', it needs the 'targets = batch[-1]'

**Comment by DonkeyShot21** [comment]
Yes it needs the targets to compute the nn accuracy (line 127 in `nnclr.py`). If you do not want to log the nn accuracy you can just remove the attribute `self.queue_y` from the class. Hope it helps.

**Comment by HBU-Lin-Li** [comment]
when i use 'find_nn', and  I  do this:''idx1, nn1 = find_nn(z_i)          _, nn2 = find_nn(z_j)", but i get this error :find_nn() missing 1 required positional argument: 'z'.  So why?

**Comment by DonkeyShot21** [comment]
Mmmh I am not sure, test are passing so it might be a problem with your implementation. Did you remember to remove `self` if you are not using `find_nn` inside a class?

**Comment by vturrisi** [comment]
Hey. Just to extend on what @DonkeyShot21 said, if you want to reuse the queue functions without y, you can probably do the following:

1- add this to your `__init__`:
```
self.register_buffer("queue", torch.randn(self.queue_size, output_dim))
self.queue = F.normalize(self.queue, dim=1)
self.register_buffer("queue_ptr", torch.zeros(1, dtype=torch.long))
```
2- create `dequeue_and_enqueue` without `y` in it:
```
@torch.no_grad()
def find_nn(self, z):
    idx = (z @ self.queue.T).max(dim=1)[1]
    nn = self.queue[idx]
    return idx, nn
```
3- copy `find_nn`:
```
@torch.no_grad()
def dequeue_and_enqueue(self, z):
   z = gather(z)
   batch_size = z.shape[0]

   ptr = int(self.queue_ptr)
   assert self.queue_size % batch_size == 0

   self.queue[ptr : ptr + batch_size, :] = z
   ptr = (ptr + batch_size) % self.queue_size

   self.queue_ptr[0] = ptr
```
Note: also remember to remove self if taking the methods outside of some class.

@DonkeyShot21 Do you think it would be good to move all the queue logic to a separate class?

**Comment by HBU-Lin-Li** [comment]
Thanks，
I have changed the function and I also get the error: find_nn() missing 1 required positional argument: 'z'.

I used such as :

 idx1, nn1 = network.Network.find_nn(z_i)
        _, nn2 = network.Network.find_nn(z_j)

**Comment by HBU-Lin-Li** [comment]
I  only want to get the 'nn' to construct contrastive loss

**Comment by vturrisi** [comment]
I think you are calling the class method directly, so you are missing the object in `self`. Since `find_nn` requires a queue that belongs to that object, you can't call it a class method.

If you do like this and then call like `find_nn(z, queue)`, it should work:
```
@torch.no_grad()
def find_nn(z, queue):
    idx = (z @ queue.T).max(dim=1)[1]
    nn = queue[idx]
    return idx, nn
```

**Comment by HBU-Lin-Li** [comment]
Emmm, i only find that 'self.queue = F.normalize(self.queue, dim=1)', if I want to use find_nn(z, queue) in other class, how can I define queue?

**Comment by vturrisi** [comment]
Since `self.queue` is not a learnable parameter, pytorch says to define like this:
`self.register_buffer("queue", torch.randn(self.queue_size, output_dim))`

You can check more about this here: https://pytorch.org/docs/stable/generated/torch.nn.Module.html

**Comment by HBU-Lin-Li** [comment]
Thanks for your help.
what's the mean of 'queue_ptr'?

**Comment by HBU-Lin-Li** [comment]
Thanks for your help.
what's the mean of 'queue_ptr'? Targets?

**Comment by vturrisi** [comment]
No worries.
Since the queue is a matrix, we need to keep track of where we need to store elements next. For example, let's say that we have a batch of size 4. `queue_ptr` will start with a value of 0. Then, we will add the first batch elements into positions [0, 1, 2, 3]. After that, `queue_ptr` will be 4. For the second batch, we will add its instances into positions [4, 5, 6, 7] and so forth.

**Comment by HBU-Lin-Li** [comment]
Every time when I'm calling the class method directily in other  class, I always get the error: "dequeue_and_enqueue() missing 1 required positional argument: 'z'".

**Comment by vturrisi** [comment]
Yes, this is because you are actually missing `self` if you wanna call it as a class method, you need to add a @classmethod decorator and fix the need for self.something inside the methods

**Comment by vturrisi** [comment]
I'll close the issue, but feel free to open if you need anything :)
